File: opensd/pipe.py
import math
import numpy as np

class Pipe(object):

    def __init__(self,identifier,circuit,diameter,length,unode,ufrac,dnode,dfrac,fricopt,roughness,ncell,heat_input,cfarea,npar,qcrit,Kforward,flowreg):
        self.identifier=identifier
        self.circuit = circuit
        self.flowreg = flowreg

        self.diameter=diameter
        self.length=length

        self.ncell=ncell
        self.unode=unode
        self.ufrac = ufrac
        self.dnode=dnode
        self.dfrac = dfrac
        
        self.npar = npar
        self.qcrit = qcrit
        self.Kforward_old = self.Kforward = Kforward

        delx = length/ncell
        delz = (dnode.elevation - unode.elevation)/ncell
        if delz - delx > 1.E-6:
            print ("length is less than the elevation difference. stopping",self.identifier)
            sys.exit()
        
        self.cfarea1 = cfarea
        if cfarea is None:
            cfarea=math.pi*diameter**2/4.
            
        cfarea = cfarea * npar
            
        self.heat_input = heat_input

        self.nodes = []
        for i in range(ncell-1):
            node=circuit.add_node(identifier+"_node"+str(i),volume=delx*cfarea,elevation=unode.elevation+(dnode.elevation-unode.elevation)*(i+1)/ncell)
            node.i = i
            self.nodes.append(node)
        
        self.faces = []
        for i in range(ncell):
            if i==0 and ncell==1:
                self.faces.append(PFace(i,self,unode,ufrac,dnode,dfrac,diameter,cfarea,delx,delz,fricopt,roughness))
            elif i==0:
                self.faces.append(PFace(i,self,unode,ufrac,self.nodes[0],None,diameter,cfarea,delx,delz,fricopt,roughness))
            elif i==ncell-1:
                self.faces.append(PFace(i,self,self.nodes[ncell-2],None,dnode,dfrac,diameter,cfarea,delx,delz,fricopt,roughness))
            else:
                self.faces.append(PFace(i,self,self.nodes[i-1],None,self.nodes[i],None,diameter,cfarea,delx,delz,fricopt,roughness))
            
        for i in range(ncell-1):
            self.nodes[i].ifaces.append(self.faces[i])
            self.nodes[i].ofaces.append(self.faces[i+1])

        unode.ofaces.append(self.faces[0])
        unode.volume = unode.volume + 0.5*delx*cfarea

        dnode.ifaces.append(self.faces[-1])
        dnode.volume = dnode.volume + 0.5*delx*cfarea
        
        self.mflow = 0.
        
    # def add_wall(self,thk,solname,sollib,restraint):
        # wall = Wall(thk,solname,sollib,restraint)
        # for face in self.faces:
            # face.wall = wall
    
class Face(object): #partial class
    def __init__(self,faceno,unode,ufrac,dnode,dfrac):
        
        self.vflow_old=np.array(1.E-8)
        self.mflow = 0.
        self.velocity=0.

        self.unode = unode
        self.ufrac = ufrac
        if ufrac is not None and isinstance(self.unode,Reservoir): self.uheight = ufrac*self.unode.height
        self.dnode = dnode
        self.dfrac = dfrac
        if dfrac is not None and isinstance(self.dnode,Reservoir): self.dheight = dfrac*self.dnode.height
        
        
        # self.heat_input = self.heat_input_old = 0.
        self.heat_input_old = 0.
        self.heat_input = 0.
        self.heat_hslab = self.heat_hslab_old = []
        self.choked = False
        self.presidue = 0.
        self.Gcr = 1.E8
        self.pcr = 0.

    def assign_statevar(self):
        self.tpres_old = 0.5*(self.unode.tpres_old+self.dnode.tpres_old)
        self.spres_old = 0.5*(self.unode.spres_old+self.dnode.spres_old)
        self.ttemp_old = 0.5*(self.unode.ttemp_old+self.dnode.ttemp_old)
        self.stemp_old = 0.5*(self.unode.stemp_old+self.dnode.stemp_old)

class PFace(Face):

    def __init__(self,faceno,pipe,unode,ufrac,dnode,dfrac,diameter,cfarea,delx,delz,fricopt,roughness):
        super().__init__(faceno,unode,ufrac,dnode,dfrac)
        self.circuit = pipe.circuit
        self.faceno=faceno
        self.pipe=pipe
        self.diameter=diameter
        self.cfarea=cfarea
        self.delx=delx
        self.delz=delz
        self.roughness=roughness
        self.Re = 0. 
        self.fricopt = fricopt
        self.fricfact_old = 64. #maximum initial friction factor considered
        # self.flstate = self.pipe.circuit.flstate
        self.opening = 1.
        self.circuit.faces.append(self)

File: src/pipe.cpp
//! \file pipe.cpp
#include "opensd/pipe.h"

#include <iostream>

#include "opensd/error.h"
#include "opensd/xml_interface.h"
#include "opensd/vector.h"

namespace opensd {

//==============================================================================
// Global variables
//==============================================================================

//==============================================================================
// Pipe implementation
//==============================================================================

Pipe::Pipe(pugi::xml_node pipe_node)
{
  if (check_for_node(pipe_node, "identifier")) {
    this->identifier = get_node_value(pipe_node, "identifier");
  } else {
    fatal_error("Must specify identifier of pipe in geometry XML file.");
  }

  this->diameter  = stod(get_node_value(pipe_node, "diameter"));
  this->length    = stod(get_node_value(pipe_node, "length"));
  this->dnode_str = get_node_value(pipe_node, "dnode");
  this->unode_str = get_node_value(pipe_node, "unode");
  this->ncell     = 1; //stod(get_node_value(pipe_node, "ncell"));
  this->unode = nullptr;
  this->dnode = nullptr;
  double ufrac;
  double dfrac;
  double cfarea;
  double delx;
  double delz;
  this->roughness = stod(get_node_value(pipe_node, "roughness"));
  double fricopt;

}


/*Pipe::Pipe(std::string identifier, Circuit* circuit, double diameter, double length, Node* unode, double ufrac, Node* dnode, double dfrac, double ficopt, double roughness, int ncell, double heat_input, double cfarea, int npar, double qcrit, double Kforward, int flowreg)
    : identifier(identifier), circuit(circuit), diameter(diameter), length(length), unode(unode), ufrac(ufrac), dnode(dnode), dfrac(dfrac), ncell(ncell), heat_input(heat_input), npar(npar), qcrit(qcrit), Kforward_old(Kforward), Kforward(Kforward), cfarea1(cfarea1) {

     double delx = length / ncell;
    double delz = (dnode.elevation - unode.elevation) / ncell;

    if (delz - delx > 1.E-6) {
        std::cout << "length is less than the elevation difference. stopping " << identifier << std::endl;
        exit(EXIT_FAILURE);
    }

    if (cfarea == NULL) {
        cfarea = M_PI * std::pow(diameter, 2) / 4.0;
    }

    cfarea *= npar;

    nodes.reserve(ncell - 1);
    for (int i = 0; i < ncell - 1; ++i) {
        Node node(identifier + "_node" + std::to_string(i), delx * cfarea, unode.elevation + (dnode.elevation - unode.elevation) * (i + 1) / ncell);
        node.i = i;
        nodes.push_back(node);
    }

    faces.reserve(ncell);
    for (int i = 0; i < ncell; ++i) {
        if (i == 0 && ncell == 1) {
            faces.push_back(PFace(i, *this, unode, ufrac, dnode, dfrac, diameter, cfarea, delx, delz, ficopt, roughness));
        } else if (i == 0) {
            faces.push_back(PFace(i, *this, unode, ufrac, nodes[0], NULL, diameter, cfarea, delx, delz, ficopt, roughness));
        } else if (i == ncell - 1) {
            faces.push_back(PFace(i, *this, nodes[ncell - 2], NULL, dnode, dfrac, diameter, cfarea, delx, delz, ficopt, roughness));
        } else {
            faces.push_back(PFace(i, *this, nodes[i - 1], NULL, nodes[i], NULL, diameter, cfarea, delx, delz, ficopt, roughness));
        }
    }

    for (int i = 0; i < ncell - 1; ++i) {
        nodes[i].ifaces.push_back(faces[i]);
        nodes[i].ofaces.push_back(faces[i + 1]);
    }

    unode.ofaces.push_back(faces[0]);
    unode.volume += 0.5 * delx * cfarea;

    dnode.ifaces.push_back(faces[ncell - 1]);
    dnode.volume += 0.5 * delx * cfarea;

    mflow = 0.0;
 }*/

// void Pipe::add_wall(double thk, std::string solname, std::string sollib, int restraint) {
    // Wall wall(thk, solname, sollib, restraint);
    // for (auto& face : faces) {
        // face.wall = wall;
    // }
// }

// void Pipe::update_mflow() {
    // std::vector<double> mface;
    // for (auto& face : faces) {
        // mface.push_back(face.mflow);
    // }
    // if (!mface.empty()) {
        // mflow = std::accumulate(mface.begin(), mface.end(), 0.0) / mface.size();
    // }
// }



} // namespace opensd
